File: C++/1_Creational Patterns/4_Builder/Pizza_Order.cpp
#include <iostream>
#include <vector>
#include <string>

class Pizza {
public:
    class PizzaBuilder;

private:
    std::string size;
    std::string crust;
    std::string cheese;
    std::vector<std::string> toppings;

    Pizza(PizzaBuilder* builder);

public:
    class PizzaBuilder {
        friend class Pizza;
    private:
        std::string size;
        std::string crust;
        std::string cheese;
        std::vector<std::string> toppings;

    public:
        PizzaBuilder(std::string size): size(size) {}

        PizzaBuilder& crust(std::string crust) {
            this->crust = crust;
            return *this;
        }

        PizzaBuilder& cheese(std::string cheese) {
            this->cheese = cheese;
            return *this;
        }

        PizzaBuilder& addTopping(std::string topping) {
            this->toppings.push_back(topping);
            return *this;
        }

        Pizza* build() {
            return new Pizza(this);
        }
    };

    void print();
};

Pizza::Pizza(PizzaBuilder* builder): 
    size(builder->size), crust(builder->crust), cheese(builder->cheese), toppings(builder->toppings) {}

void Pizza::print() {
    std::cout << "Pizza of size " << size << ", crust " << crust << ", cheese " << cheese << " with toppings: ";
    for (std::string topping : toppings) {
        std::cout << topping << " ";
    }
    std::cout << "\n";
}

int main() {
    Pizza::PizzaBuilder builder("Large");
    Pizza* pizza = builder.crust("Thin")
        .cheese("Mozzarella")
        .addTopping("Pepperoni")
        .addTopping("Mushrooms")
        .addTopping("Onions")
        .build();

    pizza->print();
    delete pizza;
}

File: C#/1_Creational Patterns/4_Builder/Pizza_Order.cs
using System;
using System.Collections.Generic;

public class Pizza
{
    public string Size { get; private set; }
    public string Crust { get; private set; }
    public string Cheese { get; private set; }
    public List<string> Toppings { get; private set; }

    private Pizza(PizzaBuilder builder)
    {
        Size = builder.Size;
        Crust = builder.Crust;
        Cheese = builder.Cheese;
        Toppings = builder.Toppings;
    }

    public class PizzaBuilder
    {
        public string Size { get; private set; }
        public string Crust { get; set; }
        public string Cheese { get; set; }
        public List<string> Toppings { get; private set; } = new List<string>();

        public PizzaBuilder(string size)
        {
            Size = size;
        }

        public PizzaBuilder SetCrust(string crust)
        {
            Crust = crust;
            return this;
        }

        public PizzaBuilder SetCheese(string cheese)
        {
            Cheese = cheese;
            return this;
        }

        public PizzaBuilder AddTopping(string topping)
        {
            Toppings.Add(topping);
            return this;
        }

        public Pizza Build()
        {
            return new Pizza(this);
        }
    }
}

class Program
{
    static void Main(string[] args)
    {
        var pizza = new Pizza.PizzaBuilder("Large")
            .SetCrust("Thin")
            .SetCheese("Mozzarella")
            .AddTopping("Pepperoni")
            .AddTopping("Mushrooms")
            .AddTopping("Onions")
            .Build();

        Console.WriteLine(pizza);
    }
}
